lms/djangoapps/courseware/views/views.py

```diff
@@ -749,40 +743,121 @@ def _progress(request, course_key, student_id):
     }
 
-    if show_generate_cert_btn:
-        # If current certificate is invalidated by instructor
-        # then show the certificate invalidated message.
-        context.update({
-            'certificate_invalidated': certs_api.is_certificate_invalid(student, course_key)
-        })
+    with outer_atomic():
+        response = render_to_response('courseware/progress.html', context)
+
+    return response
+
+
+def _get_cert_data(student, course, course_key, is_active, enrollment_mode):
+    """Returns students course certificate related data.
+
+    Arguments:
+        student (User): Student for whom certificate to retrieve.
+        course (Course): Course object for which certificate data to retrieve.
+        course_key (CourseKey): Course identifier for course.
+        is_active (Bool): Boolean value to check if course is active.
+        enrollment_mode (String): Course mode in which student is enrolled.
 
-        cert_status = certs_api.certificate_downloadable_status(student, course_key)
-        context.update(cert_status)
-        # showing the certificate web view button if feature flags are enabled.
+    Returns:
+        returns dict if course certificate is available else None.
+    """
+    show_generate_cert_btn = (
+        is_active and CourseMode.is_eligible_for_certificate(enrollment_mode)
+        and certs_api.cert_generation_enabled(course_key)
+    )
+
+    if enrollment_mode == CourseMode.AUDIT:
+        return CertData(
+            CertificateStatuses.audit_passing,
+            'Your enrollment: Audit track',
+            'You are enrolled in the audit track for this course. The audit track does not include a certificate.',
+            download_url=None,
+            show_cert_web_view=False,
+            cert_web_view_url=None
+        )
+
+    if not show_generate_cert_btn:
+        return None
+
+    if certs_api.is_certificate_invalid(student, course_key):
+        return CertData(
+            CertificateStatuses.invalidated,
+            'Your certificate has been invalidated',
+            'Please contact your course team if you have any questions.',
+            download_url=None,
+            show_cert_web_view=False,
+            cert_web_view_url=None
+        )
+
+    cert_downloadable_status = certs_api.certificate_downloadable_status(student, course_key)
+    # If the learner is in verified modes and the student did not have
+    # their ID verified, we need to show message to ask learner to verify their ID first
+    missing_required_verification = enrollment_mode in CourseMode.VERIFIED_MODES and \
+        not SoftwareSecurePhotoVerification.user_is_verified(student)
+
+    if cert_downloadable_status['is_downloadable']:
+        cert_status = CertificateStatuses.downloadable
+        title = 'Your certificate is available'
+        msg = 'You can keep working for a higher grade, or request your certificate now.'
+        download_url = cert_downloadable_status['download_url']
```

Comment:
`download_url` is set here, but only used in the last fallthrough case. Given that, it would be clearer to simply set it down there, right before it's used, or to not even store it as a separate variable.